keyboards/mouse/ploopyco_trackball/ploopyco_trackball.c

```diff
@@ -0,0 +1,247 @@
+    // Lovingly ripped from the Ploopy Source
+
+    // If the mouse wheel was just released, do not scroll.
+    // unsigned long elapsed = micros() - middleClickRelease;
+    // if (elapsed < SCROLL_BUTT_DEBOUNCE) { return 0; }
+
+    // Limit the number of scrolls per unit time.
+    if ((timer_read() - lastScroll) < OPT_DEBOUNCE) return;
+
+    // Don't scroll if the middle button is depressed.
+    // int middleButtonPin = digitalRead(MOUSE_MIDDLE_PIN);
+    // if (middleButtonPin == LOW) { return 0; }
+
+    lastScroll  = timer_read();
+    uint16_t p1 = adc_read(OPT_ENC1_MUX);
+    uint16_t p2 = adc_read(OPT_ENC2_MUX);
+    if (DEBUGOPTO) dprintf("OPT1: %d, OPT2: %d\n", p1, p2);
+
+    uint8_t dir = 0;
+    if (p1 < OPT_THRES && p2 < OPT_THRES) {
+        if (OptLowPin == OPT_ENC1) {
+            dir = -1;
+        }  // scroll down
+        else if (OptLowPin == OPT_ENC2) {
+            dir = 1;
+        }  // scroll up
+        OptLowPin = 0;
+    } else if (p1 < OPT_THRES) {
+        OptLowPin = OPT_ENC1;
+    } else if (p2 < OPT_THRES) {
+        OptLowPin = OPT_ENC2;
+    }
+
+    // Bundle and send if needed
+    if (dir == 0) return;
+    report_mouse_t cRep = pointing_device_get_report();
+    cRep.v += dir * OPT_SCALE;
+    pointing_device_set_report(cRep);
+}
+
+// Hardware Setup
+void keyboard_pre_init_kb(void) {
+    // debug_enable = false;
+    // debug_matrix = true;
+    // debug_mouse  = false;
+
+    // Set up all the hardware
+    setPinOutput(SENSOR_CS);
+
+    // These should probably be moved into the matrix itself
+    // using DIRECT_PINS, and then custom keycodes added for the
+    // mouse buttons.
+    setPinInputHigh(MOUSE_LEFT_PIN);
+    setPinInputHigh(MOUSE_RIGHT_PIN);
+    setPinInputHigh(MOUSE_MIDDLE_PIN);
+    setPinInputHigh(MOUSE_BACK_PIN);
+    setPinInputHigh(MOUSE_FORWARD_PIN);
+
+    // This can probably be replaced with rotary encoder config,
+    setPinInput(OPT_ENC1);
+    setPinInput(OPT_ENC2);
+
+    // This is the debug LED.
+    setPinOutput(F7);
+    if (debug_enable) {
+        writePinHigh(F7);
+    } else {
+        writePinLow(F7);
+    }
+
+    /* Ground all output pins connected to ground. This provides additional
+     * pathways to ground. If you're messing with this, know this: driving ANY
+     * of these pins high will cause a short. On the MCU. Ka-blooey.
+     */
+    setPinOutput(D3);
+    writePinLow(D3);
+    setPinOutput(D1);
+    writePinLow(D1);
+    setPinOutput(B4);
+    writePinLow(B4);
+    setPinOutput(B6);
+    writePinLow(B6);
+    setPinOutput(B7);
+    writePinLow(B7);
+    setPinOutput(D6);
+    writePinLow(D6);
+    setPinOutput(C7);
+    writePinLow(C7);
+    setPinOutput(F6);
+    writePinLow(F6);
+    setPinOutput(F5);
+    writePinLow(F5);
+    setPinOutput(F3);
+    writePinLow(F3);
+
+    // Initialize SPI for MCU
+    SPI_Init(SPI_OPTION);
+    DDRB |= (1 << SS_TB);
+    PORTB |= (1 << SS_TB);  // pull up to diable all comm
```

Comment:
We should convert this over to `spi_master`.

Reply:
We should, but looking at it, I'm ... not exactly sure how to do so.